From 015c8551c8166d0a3eb3d615adcbacc3da702109 Mon Sep 17 00:00:00 2001
From: Yong Zhao <Yong.Zhao@amd.com>
Date: Mon, 17 Apr 2017 15:27:20 -0400
Subject: [PATCH 2018/2094] drm/amdgpu: Revision for "remove the unused
 vm_manager lock"

This commit improves the changes made in
commit 540cd761f30530f1b8aa2ee7e9920ebb0b7c1d1e
Author: Evan Quan <evan.quan@amd.com>
Date:   Thu Apr 13 14:43:58 2017 +0800

    drm/amdgpu: remove the unused vm_manager lock
    - the vm_manager lock was removed by "drm/amdgpu: split
      VMID management by VMHUB" - fine granularity lock was introduced stead

Change-Id: I89c164f0957887617bfdd3b03c2d126810a4d27c
Signed-off-by: Yong Zhao <Yong.Zhao@amd.com>
---
 drivers/gpu/drm/amd/amdgpu/amdgpu_vm.c | 10 ++++++++++
 1 file changed, 10 insertions(+)

diff --git a/drivers/gpu/drm/amd/amdgpu/amdgpu_vm.c b/drivers/gpu/drm/amd/amdgpu/amdgpu_vm.c
index a693e96..cc773cc 100644
--- a/drivers/gpu/drm/amd/amdgpu/amdgpu_vm.c
+++ b/drivers/gpu/drm/amd/amdgpu/amdgpu_vm.c
@@ -2352,6 +2352,10 @@ int amdgpu_vm_init(struct amdgpu_device *adev, struct amdgpu_vm *vm,
 
 	vm->is_kfd_vm = is_kfd_vm;
 	if (is_kfd_vm) {
+		struct amdgpu_vm_id_manager *id_mgr =
+				&adev->vm_manager.id_mgr[AMDGPU_GFXHUB];
+
+		mutex_lock(&id_mgr->lock);
 		if (adev->vm_manager.n_kfd_vms++ == 0) {
 			/* First KFD VM: enable compute power profile */
 			if (adev->pp_enabled)
@@ -2361,6 +2365,7 @@ int amdgpu_vm_init(struct amdgpu_device *adev, struct amdgpu_vm *vm,
 				adev->pm.funcs->switch_power_profile(adev,
 						AMD_PP_COMPUTE_PROFILE);
 		}
+		mutex_unlock(&id_mgr->lock);
 	}
 
 	return 0;
@@ -2415,6 +2420,10 @@ void amdgpu_vm_fini(struct amdgpu_device *adev, struct amdgpu_vm *vm)
 	int i;
 
 	if (vm->is_kfd_vm) {
+		struct amdgpu_vm_id_manager *id_mgr =
+				&adev->vm_manager.id_mgr[AMDGPU_GFXHUB];
+
+		mutex_lock(&id_mgr->lock);
 		WARN(adev->vm_manager.n_kfd_vms == 0, "Unbalanced number of KFD VMs");
 
 		if (--adev->vm_manager.n_kfd_vms == 0) {
@@ -2426,6 +2435,7 @@ void amdgpu_vm_fini(struct amdgpu_device *adev, struct amdgpu_vm *vm)
 				adev->pm.funcs->switch_power_profile(adev,
 						AMD_PP_GFX_PROFILE);
 		}
+		mutex_unlock(&id_mgr->lock);
 	}
 
 	amd_sched_entity_fini(vm->entity.sched, &vm->entity);
-- 
2.7.4

